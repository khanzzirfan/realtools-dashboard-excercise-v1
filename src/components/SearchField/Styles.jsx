import styled from 'styled-components';
import TextField from 'components/Form/TextField';
import { ReactComponent as IconSearch } from 'theme/icons/search.svg';

const rem = (size) => `${parseInt(size) / 16}rem`;

export const SearchInput = styled(TextField)`
  margin-left: 12px;
  margin-right: 6px;
  margin-top: 12px;
  margin-bottom: 11px;
  border: none;
  font-size: ${rem(14)};
  line-height: 1.25;
  padding: 0;
`;

export const StyledIconSearch = styled(IconSearch)`
  height: 16px;
  width: 16px;
  & > path {
    fill: var(--colors-lynch);
  }
`;

export const SearchFieldWrapper = styled('div')`
  display: flex;
  align-items: center;

  ${({ filled }) =>
    filled &&
    `background: var(----colors-zircon);
    border: 1px solid var(--colors-selago);
    box-shadow: inset 1px 2px 3px var(--colors-lightspindle);
    border-radius: 20px;
    `};

  &:hover {
    border: 1px solid var(--colors-selago);
    border-radius: 20px;
    background: var(----colors-zircon);
  }
  &:active,
  &:focus {
    background: var(----colors-zircon);
    border: 1px solid var(--colors-selago);
    box-shadow: inset 1px 2px 3px var(--colors-lightspindle);
    border-radius: 20px;
  }
  padding-right: 24px;
  padding-left: 24px;
`;
