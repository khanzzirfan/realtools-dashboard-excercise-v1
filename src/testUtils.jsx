import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { QueryClient, QueryClientProvider } from 'react-query';
import { MemoryRouter } from 'react-router-dom';
import { AppContextProvider } from './hooks/useAppContext';
import { theme } from './theme';

const AllTheProviders = ({ children }) => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        // âœ… turns retries off
        retry: 2,
      },
    },
  });
  return (
    <QueryClientProvider client={queryClient}>
      <AppContextProvider>
        <ThemeProvider theme={theme}>
          <MemoryRouter initialEntries={['/people', '/people/new', '/people/edit/:id']}>
            {children}
          </MemoryRouter>
        </ThemeProvider>
      </AppContextProvider>
    </QueryClientProvider>
  );
};

const customRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
