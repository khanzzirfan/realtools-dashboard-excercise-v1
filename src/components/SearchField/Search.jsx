import { SearchFieldWrapper, StyledIconSearch, SearchInput } from './Styles';
import PropTypes from 'prop-types';

export default function Search({ placeholder, value, ...props }) {
  return (
    <SearchFieldWrapper filled={!!value}>
      <StyledIconSearch />
      <SearchInput
        type="search"
        id={'search'}
        aria-label="Search for employees"
        label=""
        placeholder={placeholder}
        value={value}
        {...props}
      />
    </SearchFieldWrapper>
  );
}

Search.propsTypes = {
  placeholder: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  onChange: PropTypes.func.isRequired,
};
