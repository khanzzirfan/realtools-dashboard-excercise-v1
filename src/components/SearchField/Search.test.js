import '@testing-library/jest-dom';
import { axe } from 'jest-axe';
import { render, screen, fireEvent } from '../../testUtils';
import SearchField from './Search';

describe('SearchField', () => {
  it('should renders correctly with default styles', () => {
    const onChangeFn = jest.fn();
    const { asFragment } = render(
      <SearchField placeholder="search employees" value="" onChange={onChangeFn} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render filled css styles', () => {
    const onChangeFn = jest.fn();
    const { asFragment } = render(
      <SearchField placeholder="search employees" value="Ann" onChange={onChangeFn} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should trigger onChange callback', () => {
    const onChangeFn = jest.fn();
    render(<SearchField placeholder="search employees" value="" onChange={onChangeFn} />);
    const node = screen.getByRole('searchbox');
    fireEvent.change(node, { target: { value: 'Ann' } });
    expect(onChangeFn).toHaveBeenCalledTimes(1);
  });

  /** axe voilations check */
  it('should not have axe voilations', async () => {
    const onChangeFn = jest.fn();
    const { container } = render(
      <SearchField placeholder="search employees" value="" onChange={onChangeFn} />
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
