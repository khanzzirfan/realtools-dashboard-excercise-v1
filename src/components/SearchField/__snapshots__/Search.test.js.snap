// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchField should render filled css styles 1`] = `
<DocumentFragment>
  .c3 {
  display: block;
  font-size: 0.8125rem;
  --labelColor: var(--colors-bayoux);
}

.c3:focus-within {
  --labelColor: var(--colors-irisBlue);
}

.c2 + .c2 {
  margin-top: 32px;
}

.c4 {
  display: block;
  color: var(--labelColor);
}

.c5 {
  display: block;
  width: 100%;
  background: none;
  border: none;
  border-bottom: 1.5px solid var(--colors-pigeon);
  color: var(--colors-darkBlue);
  font-size: 1.25rem;
  line-height: 1.2;
  padding: 4px 0 8px;
}

.c5:hover,
.c5:focus {
  outline: none;
  border-bottom-color: var(--colors-irisBlue);
}

.c5::-webkit-input-placeholder,
.c5::-webkit-input-placeholder {
  color: var(--colors-bayoux);
}

.c5::-webkit-input-placeholder,
.c5::-moz-placeholder {
  color: var(--colors-bayoux);
}

.c5::-webkit-input-placeholder,
.c5:-ms-input-placeholder {
  color: var(--colors-bayoux);
}

.c5::-webkit-input-placeholder,
.c5::placeholder {
  color: var(--colors-bayoux);
}

.c6 {
  margin-left: 12px;
  margin-right: 6px;
  margin-top: 12px;
  margin-bottom: 11px;
  border: none;
  font-size: 0.875rem;
  line-height: 1.25;
  padding: 0;
}

.c1 {
  height: 16px;
  width: 16px;
}

.c1 > path {
  fill: var(--colors-lynch);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: var(----colors-zircon);
  border: 1px solid var(--colors-selago);
  box-shadow: inset 1px 2px 3px var(--colors-lightspindle);
  border-radius: 20px;
  padding-right: 24px;
  padding-left: 24px;
}

.c0:hover {
  border: 1px solid var(--colors-selago);
  border-radius: 20px;
  background: var(----colors-zircon);
}

.c0:active,
.c0:focus {
  background: var(----colors-zircon);
  border: 1px solid var(--colors-selago);
  box-shadow: inset 1px 2px 3px var(--colors-lightspindle);
  border-radius: 20px;
}

<div
    class="c0"
  >
    <svg
      class="c1"
    >
      search.svg
    </svg>
    <label
      class="c2 c3"
    >
      <span
        class="c4"
        for="search"
      />
      <input
        aria-label="Search for employees"
        class="c5 c6"
        id="search"
        placeholder="search employees"
        type="search"
        value="Ann"
      />
    </label>
  </div>
</DocumentFragment>
`;

exports[`SearchField should renders correctly with default styles 1`] = `
<DocumentFragment>
  .c3 {
  display: block;
  font-size: 0.8125rem;
  --labelColor: var(--colors-bayoux);
}

.c3:focus-within {
  --labelColor: var(--colors-irisBlue);
}

.c2 + .c2 {
  margin-top: 32px;
}

.c4 {
  display: block;
  color: var(--labelColor);
}

.c5 {
  display: block;
  width: 100%;
  background: none;
  border: none;
  border-bottom: 1.5px solid var(--colors-pigeon);
  color: var(--colors-darkBlue);
  font-size: 1.25rem;
  line-height: 1.2;
  padding: 4px 0 8px;
}

.c5:hover,
.c5:focus {
  outline: none;
  border-bottom-color: var(--colors-irisBlue);
}

.c5::-webkit-input-placeholder,
.c5::-webkit-input-placeholder {
  color: var(--colors-bayoux);
}

.c5::-webkit-input-placeholder,
.c5::-moz-placeholder {
  color: var(--colors-bayoux);
}

.c5::-webkit-input-placeholder,
.c5:-ms-input-placeholder {
  color: var(--colors-bayoux);
}

.c5::-webkit-input-placeholder,
.c5::placeholder {
  color: var(--colors-bayoux);
}

.c6 {
  margin-left: 12px;
  margin-right: 6px;
  margin-top: 12px;
  margin-bottom: 11px;
  border: none;
  font-size: 0.875rem;
  line-height: 1.25;
  padding: 0;
}

.c1 {
  height: 16px;
  width: 16px;
}

.c1 > path {
  fill: var(--colors-lynch);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-right: 24px;
  padding-left: 24px;
}

.c0:hover {
  border: 1px solid var(--colors-selago);
  border-radius: 20px;
  background: var(----colors-zircon);
}

.c0:active,
.c0:focus {
  background: var(----colors-zircon);
  border: 1px solid var(--colors-selago);
  box-shadow: inset 1px 2px 3px var(--colors-lightspindle);
  border-radius: 20px;
}

<div
    class="c0"
  >
    <svg
      class="c1"
    >
      search.svg
    </svg>
    <label
      class="c2 c3"
    >
      <span
        class="c4"
        for="search"
      />
      <input
        aria-label="Search for employees"
        class="c5 c6"
        id="search"
        placeholder="search employees"
        type="search"
        value=""
      />
    </label>
  </div>
</DocumentFragment>
`;
